Different test failure reporting between 1.40 and 1.41

Hi, I noticed some differing behavior between Deno 1.40 and 1.41 around tests involving I/O. I have a test that opens a file, does some assertions, and then tries to close the file. If an assertion fails in the middle of the test, the test runner would indicate the assertion failure in 1.40, but in 1.41 it does not. Without indicating that the assertion failed (causing the test to abort, leaving the file open), it is more difficult to troubleshoot a failed test.

Sample test:
```
import { assertEquals } from "https://deno.land/std@0.219.0/assert/mod.ts";

Deno.test("always fails", async () => {
  const file = await Deno.create("something.txt");
  assertEquals(1, 2);
  file.close();
});

```

1.40
```
❯ deno --version
deno 1.40.0 (release, aarch64-apple-darwin)
v8 12.1.285.6
typescript 5.3.3
~/programming/deno-ex                                                                                                                                       15:13:34
❯ deno test -A
running 1 test from ./test.ts
always fails ... FAILED (2ms)

 ERRORS

always fails => ./test.ts:3:6
error: AssertionError: Values are not equal.


    [Diff] Actual / Expected


-   1
+   2

  throw new AssertionError(message);
        ^
    at assertEquals (https://deno.land/std@0.219.0/assert/assert_equals.ts:52:9)
    at file:///Users/danclark/programming/deno-ex/test.ts:5:3

 FAILURES

always fails => ./test.ts:3:6

FAILED | 0 passed | 1 failed (3ms)

error: Test failed
```

1.41
```
❯ deno --version
deno 1.41.2 (release, aarch64-apple-darwin)
v8 12.1.285.27
typescript 5.3.3
❯ deno test -A
Check file:///Users/danclark/programming/deno-ex/test.ts
running 1 test from ./test.ts
always fails ... FAILED (2ms)

 ERRORS

always fails => ./test.ts:3:6
error: Leaks detected:
  - A file was opened during the test, but not closed during the test. Close the file handle by calling `file.close()`.

 FAILURES

always fails => ./test.ts:3:6

FAILED | 0 passed | 1 failed (3ms)

error: Test failed
```

--- a/cli/tools/test/mod.rs
+++ b/cli/tools/test/mod.rs
@@ -792,6 +792,19 @@ async fn run_tests_for_worker_inner(
       }
     };
 
+    // Check the result before we check for leaks
+    let result = {
+      let scope = &mut worker.js_runtime.handle_scope();
+      let result = v8::Local::new(scope, result);
+      serde_v8::from_v8::<TestResult>(scope, result)?
+    };
+    if matches!(result, TestResult::Failed(_)) {
+      fail_fast_tracker.add_failure();
+      let elapsed = earlier.elapsed().as_millis();
+      sender.send(TestEvent::Result(desc.id, result, elapsed as u64))?;
+      continue;
+    }
+
     // Await activity stabilization
     if let Some(diff) = wait_for_activity_to_stabilize(
       worker,
@@ -806,6 +819,7 @@ async fn run_tests_for_worker_inner(
       let (formatted, trailer_notes) = format_sanitizer_diff(diff);
       if !formatted.is_empty() {
         let failure = TestFailure::Leaked(formatted, trailer_notes);
+        fail_fast_tracker.add_failure();
         let elapsed = earlier.elapsed().as_millis();
         sender.send(TestEvent::Result(
           desc.id,
@@ -816,12 +830,6 @@ async fn run_tests_for_worker_inner(
       }
     }
 
-    let scope = &mut worker.js_runtime.handle_scope();
-    let result = v8::Local::new(scope, result);
-    let result = serde_v8::from_v8::<TestResult>(scope, result)?;
-    if matches!(result, TestResult::Failed(_)) {
-      fail_fast_tracker.add_failure();
-    }
     let elapsed = earlier.elapsed().as_millis();
     sender.send(TestEvent::Result(desc.id, result, elapsed as u64))?;
   }

--- a/tests/specs/test/sanitizer_with_error/__test__.json
+++ b/tests/specs/test/sanitizer_with_error/__test__.json
@@ -1,0 +1,5 @@
+{
+  "args": "test --quiet --reload main.js",
+  "output": "main.out",
+  "exitCode": 1
+}

--- a/tests/specs/test/sanitizer_with_error/main.js
+++ b/tests/specs/test/sanitizer_with_error/main.js
@@ -1,0 +1,6 @@
+Deno.test(async function throws() {
+  // Leak
+  setTimeout(() => {}, 60_000);
+  // But the exception should mask the leak
+  throw new Error("Throws");
+});

--- a/tests/specs/test/sanitizer_with_error/main.js
+++ b/tests/specs/test/sanitizer_with_error/main.js
@@ -1,4 +1,4 @@
-Deno.test(async function throws() {
+Deno.test(function throws() {
   // Leak
   setTimeout(() => {}, 60_000);
   // But the exception should mask the leak

--- a/tests/specs/test/sanitizer_with_error/main.out
+++ b/tests/specs/test/sanitizer_with_error/main.out
@@ -13,6 +13,6 @@ error: Error: Throws
 
 throws => ./main.js:1:6
 
-FAILED | 0 passed | 1 failed (5ms)
+FAILED | 0 passed | 1 failed ([WILDCARD])
 
 error: Test failed

--- a/tests/specs/test/sanitizer_with_error/main.out
+++ b/tests/specs/test/sanitizer_with_error/main.out
@@ -1,5 +1,5 @@
 running 1 test from ./main.js
-throws ... FAILED (1ms)
+throws ... FAILED ([WILDCARD])
 
  ERRORS 
 

--- a/tests/specs/test/sanitizer_with_error/main.out
+++ b/tests/specs/test/sanitizer_with_error/main.out
@@ -1,0 +1,18 @@
+running 1 test from ./main.js
+throws ... FAILED (1ms)
+
+ ERRORS 
+
+throws => ./main.js:1:6
+error: Error: Throws
+  throw new Error("Throws");
+        ^
+    at throws ([WILDCARD]/main.js:5:9)
+
+ FAILURES 
+
+throws => ./main.js:1:6
+
+FAILED | 0 passed | 1 failed (5ms)
+
+error: Test failed

